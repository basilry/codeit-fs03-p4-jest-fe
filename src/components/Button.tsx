'use client'

import { ButtonHTMLAttributes, ReactElement } from "react";

interface IButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  label: string;
  onClick: () => void;
  disabled?: boolean;
}

const Button = (props: IButtonProps): ReactElement => {
  const { label, onClick, disabled } = props;

  return (
    <button data-testid="custom-button" onClick={onClick} disabled={disabled}>
      {label}
    </button>
  );
};

export default Button;