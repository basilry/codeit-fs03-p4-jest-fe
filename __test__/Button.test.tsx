import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event"; // user-event 추가
import Button from "@/components/Button";

describe("🔹 Button Component Test", () => {
  test("✅ 버튼이 정상적으로 렌더링 되는지 확인", () => {
    render(<Button label="버튼" onClick={() => {}} />);
    const buttonElement = screen.getByText("버튼");
    expect(buttonElement).toBeInTheDocument();
  });

  test("✅ 버튼 클릭 이벤트가 정상적으로 동작하는지 확인 (기본)", () => {
    const handleClick = jest.fn();
    render(<Button label="버튼" onClick={handleClick} />);
    
    const buttonElement = screen.getByText("버튼");
    buttonElement.click();

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test("✅ userEvent를 사용하여 버튼 클릭", async () => {
    const handleClick = jest.fn();
    const user = userEvent.setup();
    render(<Button label="버튼" onClick={handleClick} />);
    
    const buttonElement = screen.getByRole("button", { name: "버튼" });
    await user.click(buttonElement);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test("✅ 버튼이 비활성화(disabled) 상태일 때 클릭이 안 되는지 확인", () => {
    const handleClick = jest.fn();
    render(<Button label="버튼" onClick={handleClick} disabled />);
    
    const buttonElement = screen.getByText("버튼");
    expect(buttonElement).toBeDisabled();

    buttonElement.click();
    expect(handleClick).not.toHaveBeenCalled();
  });

  test("✅ Snapshot 테스트", () => {
    const { container } = render(<Button label="스냅샷 테스트" onClick={() => {}} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});