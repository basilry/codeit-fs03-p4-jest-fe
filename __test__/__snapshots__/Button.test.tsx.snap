// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ğŸ”¹ Button Component Test âœ… Snapshot í…ŒìŠ¤íŠ¸ 1`] = `
<button
  data-testid="custom-button"
>
  ìŠ¤ëƒ…ìƒ· í…ŒìŠ¤íŠ¸
</button>
`;
